lib/cli/json_args.rs

```diff
@@ -0,0 +1,1019 @@
+
+mod error;
+pub mod help;
+
+use std::str::FromStr;
+
+use serde::{Deserialize, Serialize};
+use serde_json::{json, Value};
+
+use casper_types::{
+    bytesrepr::{ToBytes, OPTION_NONE_TAG, OPTION_SOME_TAG, RESULT_ERR_TAG, RESULT_OK_TAG},
+    AsymmetricType, CLType, CLValue, Key, NamedArg, PublicKey, URef, U128, U256, U512,
+};
+
+use crate::cli::CliError;
+pub use error::{Error, ErrorDetails};
+
+#[derive(Clone, Serialize, Deserialize, Debug)]
+pub(super) struct JsonArg {
+    name: String,
+    #[serde(rename = "type")]
+    cl_type: CLType,
+    value: Value,
+}
+
+impl TryFrom<JsonArg> for NamedArg {
+    type Error = CliError;
+
+    fn try_from(json_arg: JsonArg) -> Result<Self, Self::Error> {
+        let mut bytes = vec![];
+        json_to_bytesrepr(&json_arg.cl_type, &json_arg.value, &mut bytes).map_err(|details| {
+            Error::new(
+                json_arg.name.clone(),
+                json_arg.cl_type.clone(),
+                json_arg.value,
+                details,
+            )
+        })?;
+        Ok(NamedArg::new(
+            json_arg.name,
+            CLValue::from_components(json_arg.cl_type, bytes),
+        ))
+    }
+}
+
+fn json_to_bytesrepr(
+    cl_type: &CLType,
+    json_value: &Value,
+    output: &mut Vec<u8>,
+) -> Result<(), ErrorDetails> {
+    match (cl_type, json_value) {
+        (&CLType::Bool, Value::Bool(bool)) => bool.write_bytes(output)?,
+        (&CLType::I32, Value::Number(number)) => {
+            let value = number
+                .as_i64()
+                .and_then(|value| i32::try_from(value).ok())
+                .ok_or(ErrorDetails::CannotParseToI32)?;
+            value.write_bytes(output)?
+        }
+        (&CLType::I64, Value::Number(number)) => {
+            let value = number.as_i64().ok_or(ErrorDetails::CannotParseToI64)?;
+            value.write_bytes(output)?
+        }
+        (&CLType::U8, Value::Number(number)) => {
+            let value = number
+                .as_u64()
+                .and_then(|value| u8::try_from(value).ok())
+                .ok_or(ErrorDetails::CannotParseToU8)?;
+            value.write_bytes(output)?
+        }
+        (&CLType::U32, Value::Number(number)) => {
+            let value = number
+                .as_u64()
+                .and_then(|value| u32::try_from(value).ok())
+                .ok_or(ErrorDetails::CannotParseToU32)?;
+            value.write_bytes(output)?
+        }
+        (&CLType::U64, Value::Number(number)) => {
+            let value = number.as_u64().ok_or(ErrorDetails::CannotParseToU64)?;
+            value.write_bytes(output)?
+        }
+        (&CLType::U128, Value::String(string)) => {
+            let value = U128::from_dec_str(string)?;
+            value.write_bytes(output)?
+        }
+        (&CLType::U128, Value::Number(number)) => {
+            let value = number.as_u64().ok_or(ErrorDetails::CannotParseToU64)?;
+            U128::from(value).write_bytes(output)?
+        }
```

Comment:
This is just a convenience. As per the help docs: "CLTypes U128, U256 and U512 are represented as a JSON String of the decimal value, or can be represented as a Number if the value is not more than u64::MAX (18446744073709551615)"

So, a larger U128 would need to be passed as a String rather than a Number.

> Also, would it be worth it to add custom variants in ErrorDetails for these large ints?

I don't think so, since we have `ParseBigint` error for the case where we're parsing from a String.  If we're parsing from a Number, I think `CannotParseToU64` describes the problem correctly.